### DISCUSSION TITLE
Host / Target Compilation and Toolchains

### DISCUSSION THREAD

**Discussion by steven-bellock** [DMTF/libspdm#612] [General]
The libspdm build system currently employees `DARCH` and `DTOOLCHAIN` to specify the target machine architecture (x64, aarch64, etc), compiler executable (gcc, aarch64-linux-gnu-gcc), and which `processor_bind.h` file to `#include`. However, `DTOOLCHAIN` is slightly awkward in that, for example, when targeting aarch64 with GCC, `DTOOLCHAIN=AARCH64_GCC` is used rather than `DTOOLCHAIN=GCC`. There are additional target considerations such as baremetal and endianness. In light of that this thread is to discuss possible changes to how host, target, and toolchain are specified.

In particular CMake is able to detect the host architecture, so if the host and target are the same the user should not have to specify `DARCH`. We may also need another parameter to specify variants of an architecture's toolchain. For example aarch64 can encompass `aarch64-none-elf`, `aarch64-none-linux-gnu`, and `aarch64_be-none-linux-gnu`.

**Comment by LeeTroy** [comment]
We're working on porting libspdm to two platforms of our company, and which has integrated build environments, e.g. OpenBMC (Yocto-based) and Zephyr (West + CMake).  The integrated build environment has the ability to feed cross compiler variables `CMAKE_C_{COMPILER,AR,LINKER}` into CMake. Predefining the toolchain executable makes porting harder without patching it.

**Comment by jyao1** [comment]
@LeeTroy , thanks for the info. Do you have any proposal?

**Reply by LeeTroy** [reply]
Right now, I have a very sketchy patch for it, I'll submit a PR when I get everything working.
